toolbox/check_certificates.py

```diff
@@ -0,0 +1,392 @@
+    Extract the certificate path from a configuration file.
+    :param config_path: The path to the configuration file.
+    :return: The extracted certificate path, or None if not found.
+    """
+    cert_path = None
+    try:
+        with open(config_path, 'r') as file:
+            for line in file:
+                if 'SSLCertificateFile' in line and not line.strip().startswith('#'):
+                    cert_path = line.split()[-1].strip()
+                    break
+    except Exception as e:
+        log_message(f"Error reading {config_path}: {str(e)}", error=True)
+    return cert_path
+
+
+def load_certificates(cert_path, ca_path=None):
+    """
+    Load certificates from specified paths.
+    :param cert_path: The path to the certificate file.
+    :param ca_path: The path to the CA certificate file (optional).
+    :return: A tuple of (certificate, CA certificate).
+    """
+    try:
+        cert = None
+        if cert_path:
+            with open(cert_path, 'rb') as file:
+                cert = x509.load_pem_x509_certificate(file.read(), default_backend())
+
+        ca_cert = None
+        if ca_path:
+            with open(ca_path, 'rb') as file:
+                ca_cert = x509.load_pem_x509_certificate(file.read(), default_backend())
+
+        return cert, ca_cert
+    except Exception as e:
+        log_message(
+            f"Failed to load certificates from {cert_path} or {ca_path}: {str(e)}", error=True)
+        return None, None
+
+
+def check_certificate_expiry(cert, days, cert_description):
+    """
+    Check if a certificate will expire within a specified number of days.
+    :param cert: The certificate to check.
+    :param days: The number of days to check for expiry.
+    :param cert_description: A description of the certificate.
+    """
+    if cert:
+        days_to_expire = (cert.not_valid_after - datetime.now()).days
+        message = f'{cert_description} certificate is valid for {days_to_expire} more days.'
+        log_message(message)
+        if days_to_expire <= days:
+            log_message(
+                f"Warning: The {cert_description} "
+                f"certificate will expire in {days_to_expire} days or less. "
+                f"Please renew it timely.",
+                warning=True)
+
+
+def get_certificate_from_server(server_address, port, starttls=False):
+    """
+    Retrieve the server certificate using openssl.
+    :param server_address: The address of the server.
+    :param port: The port to connect to.
+    :param starttls: Whether to use STARTTLS for LDAP connections.
+    :return: The server certificate, or None if not found.
+    """
+    try:
+        if starttls:
+            cmd = (
+                f"echo | openssl s_client -connect {server_address}:{port} -starttls ldap "
+                "2>/dev/null | openssl x509")
+        else:
+            cmd = (f"echo | openssl s_client -connect {server_address}:{port} "
+                   f"2>/dev/null | openssl x509")
```

Comment:
we could use `ssl.get_server_certificate()` here (https://docs.python.org/3.8/library/ssl.html?highlight=starttls#ssl.get_server_certificate).
This avoids the external call and we can set the TLS version.

Reply:
We could use a `socket` to get the `STARTTLS` certificate: https://stackoverflow.com/a/5655076
But we should do that later as well.